### ISSUE TITLE
feat: add support for `:scope` selector

### ISSUE THREAD

**Issue by petebacondarwin** [phenomnomnominal/tsquery#23]
See https://developer.mozilla.org/en-US/docs/Web/CSS/:scope

This selector would allow queries to limit the matching to only the direct children of the AST on which the query is run.

For example, in the following code:

```ts
const x = function() {
  function y() {
    return 'bar';
  }
  return 'foo';
};
```

How do you query for just the outer function's return statement? Currently the query `ReturnStatement` will return both return statements even if it is run on the `Block` node that is the body of the outer function.

With the `:scope` selector you would be able to use `:scope > ReturnStatement` and then run that on the body of the outer function to only return the outer `ReturnStatement`.

ESQuery has a PR for adding this already https://github.com/estools/esquery/pull/61

**Comment by petebacondarwin** [comment]
Of course since tsquery uses esquery to parse the query strings, it will be tricky to fix this until esquery supports it too...

**Comment by phenomnomnominal** [comment]
Yeah that's a great idea, and their implementation looks solid. I'll give the ESQuery team a little while to see if they respond to the comments, but then will definitely look at moving the parser into TSQuery (hopefully temporarily)

**Comment by petebacondarwin** [comment]
Great! Happy to create PR(s) as necessary. Just give me the nod.

**Comment by run1t** [comment]
It would be very helpful to be able to use the `:scope` selector.
Any news about this ?

**Comment by run1t** [comment]
I added some very basic unit tests to the PR in ESQuery, where should I do the pull requests ? 
Do I have to create another one on https://github.com/estools/esquery or do I have to pull request directly in this one: https://github.com/jupenur/esquery ? 

If you have anothers tests in mind I'll be glad to add them.
You can see my branch here: https://github.com/run1t/esquery/tree/scope-root-tests

**Comment by phenomnomnominal** [comment]
Thank you so much for your work on this! I'm not really sure how the PR should work, you could just do it against esquery and see what happens? Might be more direct that way. I guess they'll tell you if it doesn't work right.

I'd probably suggest adding another test for the use case mentioned in the original issue here as well, just to make sure we're solving the original request?

**Comment by run1t** [comment]
I'll add the test for the use case described above and I'll create a new pull request against esquery directly 😃

**Comment by run1t** [comment]
If the pull request is accepted and a new version is published, what would be the next steps (beside updating the ESquery version in package.json)?

**Comment by phenomnomnominal** [comment]
Just updating the version and then publishing it 😊

**Comment by run1t** [comment]
Unfortunately it does not seem so easy 😢 

- I published the branch on npm : https://www.npmjs.com/package/esquery-scope
- I updated the `package.json` on `tsquery``

but I have this error when I run the tests:

```
tsquery: › tsquery - :has: › should find any nodes with multiple attributes

    Unknown selector type: scope

      33 |     }
      34 |
    > 35 |     throw new Error(`Unknown selector type: ${selector.type}`);
         |           ^
      36 | }
      37 |
```
The tests below are failing with the same error:

- tsquery: › tsquery - :has: › should find any nodes with multiple attributes
- tsquery: › tsquery - :has: › should find any nodes with one of multiple attributes
- tsquery: › tsquery - :has: › should handle chained :has selectors
- tsquery: › tsquery - :has: › should handle nested :has selectors

Do you have any ideas on why it's failing ?

**Comment by phenomnomnominal** [comment]
Ah yep, too right. Once their PR gets merged and released I can port over the matcher and the tests too.

**Comment by petebacondarwin** [comment]
@run1t - I think you can test this locally more easily with `npm link` (or `yarn link`), rather than resorting to publishing your own version.

**Comment by run1t** [comment]
@petebacondarwin - Thanks for the advice !

**Comment by run1t** [comment]
I was trying to implement the [:root](https://drafts.csswg.org/selectors-4/#root-pseudo) and [:scope](https://drafts.csswg.org/selectors-4/#the-scope-pseudo) selector. I got `:root` working. But when I was working on implementing the `:scope` selector I didn't understand the differences between both.

I looked at the implementation in `ESQuery`, and they seem to differentiate the two.
https://github.com/run1t/esquery/blob/master/esquery.js#L190-#194

Do someone know the goal of this ? And why there is a scope parameter added here ?: https://github.com/run1t/esquery/blob/master/esquery.js#L50

**Comment by run1t** [comment]
@petebacondarwin Thanks for the explanation

So if we take this example:
```Typescript
function a() {
  function b() {
    return 'bar';
  }
  return 'foo';
};
```
So If I understood correctly:
###### Assuming we are in the `BLOCK` of the  `function a`:

- Executing this query `:root > FunctionDeclaration` should return  the `function a`
- Executing this query `:scope > FunctionDeclaration` should return  the `function b`

Am I right ?

@phenomnomnominal  I'll make a PR later today. Thank you !

**Comment by petebacondarwin** [comment]
That that is how I understand it. Moreover:

Assuming we are in the `BLOCK` of the `function a`:
* Executing this query `:root FunctionDeclaration` should return `function a` and `function b`
* Executing this query `:scope FunctionDeclaration` should only return  `function b`

**Comment by run1t** [comment]
@phenomnomnominal you can review the PR on #37 which should add `:scope` and `:root`
@petebacondarwin I added your use cases in the PR

**Comment by run1t** [comment]
If `esquery` doesn't want to accept the pull request how should I do ?

**Comment by phenomnomnominal** [comment]
Do we know that they don’t? Let’s give them a little nudge first, and see what needs to happen for it to be merged.

**Comment by cancerberoSgx** [comment]
If I'm understanding this correctly, I think your `:scope` it's the same as `SourceFile`. Example, `SourceFile>FunctionDeclaration` will select root function declarations while `FunctionDeclaration` will select all.

About the question, you select the outer ReturnStatement like this: `FunctionExpression >Block > * > ReturnStatement`. 

The '*' would be a SyntaxList but seems is treated specially and you cannot mention it in the query (i.e this works: `FunctionExpression >Block > SyntaxList`, this won't work: `FunctionExpression >Block > SyntaxList *`)

About `:scope` , IMO, in the context of JavaScript it could be confused with a symbol scope,

I would like to know if `SourceFile` would cover all the cases, of if I'm not understanding something about `:scope`? thanks!

**Comment by moonyoulove** [comment]
Hello, esquery has recently been updated with a feature similar to `:scope`, which is `:has(> )`. Could you please update tsquery to support this feature as well?
Here is the related commit: https://github.com/estools/esquery/commit/00448e0fcccc23e36e8ffb49bf163a7efd6ebaf7.
